### PR TITLE
zio

### PR DESCRIPTION
**Pull request by Cardu**
tested with both curl and provided client

### CODE REVIEW

File: src/main/scala/rps/GameController.scala [hunk 1]
```diff
@@ -17,11 +16,17 @@ trait GameController {
 }
 
 class GameControllerImpl(gameService: GameService)(implicit ec: ExecutionContext) extends GameController {
-  override def result(): Future[Either[Throwable, PlayResponse]] = {
-    gameService.getResult.map(p => p.flatMap(_.toRight(new IllegalStateException).map {
-      play => PlayResponse(play.userMove, play.computerMove, play.result)
-    }))
-  }
+  val runtime = Runtime.default
```
Comment (gabro) [inline src/main/scala/rps/GameController.scala]: I would get the runtime as an implicit parameter. Also I don't think you need the implicit `ExecutionContext` anymore

File: src/main/scala/rps/GameController.scala [hunk 2]
```diff
@@ -17,11 +16,17 @@ trait GameController {
...
+
+  private[this] def resultZIO = gameService
+    .getResult()
+    .map(_.getOrElse(throw new IllegalStateException))
+    .map(play => PlayResponse(play.userMove, play.computerMove, play.result))
+    .either
+
+  private[this] def playZIO(userMove: Move) = gameService.playMove(userMove).either
```
Comment (gabro) [inline src/main/scala/rps/GameController.scala]: Can you please annotate these methods? It makes it easier to understand (and this is a learning repo, so it's especially important)

File: src/main/scala/rps/GameController.scala [hunk 3]
```diff
@@ -17,11 +16,17 @@ trait GameController {
...
-  override def result(): Future[Either[Throwable, PlayResponse]] = {
-    gameService.getResult.map(p => p.flatMap(_.toRight(new IllegalStateException).map {
-      play => PlayResponse(play.userMove, play.computerMove, play.result)
-    }))
-  }
+  val runtime = Runtime.default
+
+  private[this] def resultZIO = gameService
+    .getResult()
+    .map(_.getOrElse(throw new IllegalStateException))
```
Comment (gabro) [inline src/main/scala/rps/GameController.scala]:
```suggestion
    .someOrFail(new IllegalStateException)
```

File: src/main/scala/rps/GameRepository.scala [hunk 4]
```diff
@@ -1,40 +1,43 @@
...
-import db.Tables.{Plays, PlayRow}
+import db.Tables.{PlayRow, Plays}
 import model._
+import zio.{IO, Task}
 
 trait GameRepository {
-  def save(play: Play): Future[Either[Throwable, UUID]]
-  def read(): Future[Either[Throwable, Option[Play]]]
+  def save(play: Play): IO[Throwable, UUID]
+  def read(): IO[Throwable, Option[Play]]
```
Comment (gabro) [inline src/main/scala/rps/GameRepository.scala]:
```suggestion
  def save(play: Play): Task[UUID]
  def read(): Task[Option[Play]]
```

File: src/main/scala/rps/GameRepository.scala [hunk 5]
```diff
@@ -1,40 +1,43 @@
...
 
 class GameRepositoryImpl(
   db: DatabaseDef
 )(
   implicit ec: ExecutionContext
-) extends GameRepository with SlickRepository {
+) extends GameRepository {
 
-  override def save(play: Play): Future[Either[Throwable, UUID]] = {
+  override def save(play: Play): IO[Throwable, UUID] = {
```
Comment (gabro) [inline src/main/scala/rps/GameRepository.scala]:
```suggestion
  override def save(play: Play): Task[UUID] = {
```

File: src/main/scala/rps/GameRepository.scala [hunk 6]
```diff
@@ -1,40 +1,43 @@
...
     val playRow = convertPlay(play)
     val newPlay = Plays += playRow
-    
-    futureToEither(db.run(newPlay).map(_ => playRow.id))
+
+    IO.fromFuture {
+      implicit ec => db.run(newPlay).map(_ => playRow.id)
+    }
```
Comment (gabro) [inline src/main/scala/rps/GameRepository.scala]:
```suggestion
    IO.fromFuture { _ =>
      db.run(newPlay)
    }.as(playRow.id)
```

File: src/main/scala/rps/GameRepository.scala [hunk 7]
```diff
@@ -1,40 +1,43 @@
...
   }
 
-  override def read(): Future[Either[Throwable, Option[Play]]] = { 
+  override def read(): IO[Throwable, Option[Play]] = {
```
Comment (gabro) [inline src/main/scala/rps/GameRepository.scala]:
```suggestion
  override def read(): Task[Option[Play]] = {
```

File: src/main/scala/rps/GameRepository.scala [hunk 8]
```diff
@@ -1,40 +1,43 @@
...
     val selectPlay = Plays.sortBy(_.createdAt.desc).take(1).result.headOption
-    futureToEither(db.run(selectPlay).map(_.flatMap(convertPlayRow)))
-  } 
+    val maybePlayRowIO: Task[Option[PlayRow]] = IO.fromFuture { implicit ec => db.run(selectPlay)}
```
Comment (gabro) [inline src/main/scala/rps/GameRepository.scala]:
```suggestion
    val maybePlayRowIO: Task[Option[PlayRow]] = IO.fromFuture { _ => db.run(selectPlay) }
```

File: src/main/scala/rps/GameRepository.scala [hunk 9]
```diff
@@ -1,40 +1,43 @@
...
+    maybePlayRowIO.map(_.flatMap(convertPlayRow))
+  }
 
-  private val convertPlayRow = (r: PlayRow) => for {
+  private def convertPlayRow (r: PlayRow) : Option[Play] = for {
```
Comment (gabro) [inline src/main/scala/rps/GameRepository.scala]:
```suggestion
  private def convertPlayRow(r: PlayRow): Option[Play] = for {
```

(can you use Scalafmt?)

File: src/main/scala/rps/GameService.scala [hunk 10]
```diff
@@ -1,22 +1,21 @@
...
-
 import model._
 import Move._
 import Result._
 
 trait GameService {
-  def playMove(userMove: Move): Future[Either[Throwable, UUID]]
-  def getResult(): Future[Either[Throwable, Option[Play]]]
+  def playMove(userMove: Move): IO[Throwable, UUID]
+  def getResult(): IO[Throwable, Option[Play]]
```
Comment (gabro) [inline src/main/scala/rps/GameService.scala]:
```suggestion
  def playMove(userMove: Move): Task[UUID]
  def getResult(): Task[Option[Play]]
```

File: src/main/scala/rps/GameService.scala [hunk 11]
```diff
@@ -1,22 +1,21 @@
...
 }
 
 class GameServiceImpl(repository: GameRepository) extends GameService {
-  override def playMove(userMove: Move): Future[Either[Throwable, UUID]] = {
+  override def playMove(userMove: Move): IO[Throwable, UUID] = {
```
Comment (gabro) [inline src/main/scala/rps/GameService.scala]:
```suggestion
  override def playMove(userMove: Move): Task[UUID] = {
```

File: src/main/scala/rps/GameService.scala [hunk 12]
```diff
@@ -31,6 +30,6 @@ class GameServiceImpl(repository: GameRepository) extends GameService {
   private def generateComputerMove(): Move =
     Random.shuffle(List(Rock, Paper, Scissors)).head
 
-  override def getResult(): Future[Either[Throwable, Option[Play]]] = 
-    repository.read
+  override def getResult(): IO[Throwable, Option[Play]] =
```
Comment (gabro) [inline src/main/scala/rps/GameService.scala]:
```suggestion
  override def getResult(): Task[Option[Play]] =
```

File: src/main/scala/rps/GameService.scala [hunk 13]
```diff
@@ -1,22 +1,21 @@
 package rps
 
 import java.util.UUID
-import java.time.Instant
 
-import scala.concurrent.Future
+import rps.model._
+import zio.IO
```
Comment (gabro) [inline src/main/scala/rps/GameService.scala:6]:
```suggestion
import zio.Task
```

File: .gitignore [hunk 14]
```diff
@@ -19,3 +19,9 @@ project/plugins/project/
 *.log
 
 # End of https://www.gitignore.io/api/sbt,scala
+
+### IntelliJ ###
+.idea/
+
+### Bloop ###
+.bloop/
```
Comment (gabro) [inline .gitignore]:
```suggestion
.bloop/
.metals/
project/metals.sbt
```

**Review by gabro — CHANGES_REQUESTED** [review]

File: src/main/scala/rps/GameRepository.scala (hunk 4 above)
Comment (Cardu) [reply src/main/scala/rps/GameRepository.scala]: Since this will be a tutorial, I didn't want to introduce too many names and avoided Task... also, we could put there some domain-specific errors instead of Throwable

Comment (gabro) [reply src/main/scala/rps/GameRepository.scala]: 👍 makes sense, let's then introduce a domain-specific error

File: .gitignore [hunk 15]
```diff
@@ -19,3 +19,11 @@ project/plugins/project/
 *.log
 
 # End of https://www.gitignore.io/api/sbt,scala
+
+### IntelliJ ###
+.idea/
```
Comment (veej) [inline .gitignore]: nit: in general, I think this could be in your global gitignore. AFAIK we never set IDE specific ignores.
Also, consider that is preferable to keep the diff short since these steps are meant for training. I won't add unrelated diff to it.

File: src/main/scala/rps/GameController.scala [hunk 16]
```diff
@@ -9,24 +9,23 @@ import zio._
...
-  def result(): Future[Either[Throwable, PlayResponse]]
+  def result(): Future[Either[RPSError, PlayResponse]]
 }
 
-class GameControllerImpl(gameService: GameService)(implicit ec: ExecutionContext) extends GameController {
-  val runtime = Runtime.default
+class GameControllerImpl(gameService: GameService)(implicit runtime: Runtime[ZEnv]) extends GameController {
 
-  private[this] def resultZIO = gameService
+  private[this] def resultZIO: URIO[Any, Either[RPSError, PlayResponse]] = gameService
```
Comment (gabro) [inline src/main/scala/rps/GameController.scala]:
```suggestion
  private[this] def resultZIO: UIO[Either[RPSError, PlayResponse]] = gameService
```

File: src/main/scala/rps/GameController.scala [hunk 17]
```diff
@@ -9,24 +9,23 @@ import zio._
...
     .getResult()
-    .someOrFail(new IllegalStateException)
+    .someOrFail(RPSError.NeverPlayed)
     .map(play => PlayResponse(play.userMove, play.computerMove, play.result))
     .either
 
-  private[this] def playZIO(userMove: Move) = gameService.playMove(userMove).either
+  private[this] def playZIO(userMove: Move): URIO[Any, Either[RPSError, UUID]] = gameService.playMove(userMove).either
```
Comment (gabro) [inline src/main/scala/rps/GameController.scala]:
```suggestion
  private[this] def playZIO(userMove: Move): UIO[Either[RPSError, UUID]] = gameService.playMove(userMove).either
```

**Review by gabro — CHANGES_REQUESTED** [review]

File: src/main/scala/rps/GameRepository.scala [hunk 18]
```diff
@@ -1,44 +1,49 @@
...
   }
 
-  override def read(): Future[Either[Throwable, Option[Play]]] = { 
+  override def read(): IO[RPSError, Option[Play]] = {
     val selectPlay = Plays.sortBy(_.createdAt.desc).take(1).result.headOption
-    futureToEither(db.run(selectPlay).map(_.flatMap(convertPlayRow)))
-  } 
+    val maybePlayRowIO: Task[Option[PlayRow]] = IO.fromFuture { _ => db.run(selectPlay) }
+    maybePlayRowIO.map(_.flatMap(convertPlayRow))
+  }.mapError(e => RPSError.DBError(e.getMessage))
```
Comment (gabro) [inline src/main/scala/rps/GameRepository.scala]:
```suggestion
	IO.fromFuture { _ => db.run(selectPlay) }
	  .map(_.flatMap(convertPlayRow))
	  .mapError(e => RPSError.DBError(e.getMessage)) 
  }
```

**Review by gabro — CHANGES_REQUESTED** [review]
Last pass

File: src/main/scala/rps/GameRepository.scala (hunk 18 above)
Comment (Cardu) [reply src/main/scala/rps/GameRepository.scala]: variable names were meant to help a newcomer to understand this passage deeply, step-by-step

Comment (Cardu) [reply src/main/scala/rps/GameRepository.scala]: I am the first fun of oneliners, but... I think when you are learning, even if you can read them, you don't stop to think about what's happening inside the chain

Comment (gabro) [reply src/main/scala/rps/GameRepository.scala]: ok, however I think they're a bit confusing if that's the purpose. I don't understand what `maybePlayRowIO` is by just reading the name and it feels a bit arbitrary. Anyway, I don't mind, we can leave it as is

**Review by gabro — APPROVED** [review]

File: src/main/scala/rps/GameController.scala [hunk 19]
```diff
@@ -1,27 +1,36 @@
...
-  }
+class GameControllerImpl(gameService: GameService)(implicit runtime: Runtime[ZEnv])
+    extends GameController {
+
+  private[this] def resultZIO: UIO[Either[RPSError, PlayResponse]] =
+    gameService
+      .getResult()
+      .someOrFail(RPSError.NeverPlayed)
+      .map(play => PlayResponse(play.userMove, play.computerMove, play.result))
+      .either
```
Comment (veej) [inline src/main/scala/rps/GameController.scala]: wouldn't it be better to keep the `either` outside?
I mean, in a ZIO dreamland you would like to expose `IO[E, A]` directly

File: src/main/scala/rps/GameRepository.scala (hunk 18 above)
Comment (veej) [reply src/main/scala/rps/GameRepository.scala]: I'm with @gabro. I'm not a newcomer and I struggled to understand the naming

File: src/main/scala/rps/GameRepository.scala [hunk 20]
```diff
@@ -1,44 +1,51 @@
...
+      .as(playRow.id)
+      .mapError(e => RPSError.DBError(e.getMessage))
   }
 
-  override def read(): Future[Either[Throwable, Option[Play]]] = { 
-    val selectPlay = Plays.sortBy(_.createdAt.desc).take(1).result.headOption
-    futureToEither(db.run(selectPlay).map(_.flatMap(convertPlayRow)))
-  } 
+  override def read(): IO[RPSError, Option[Play]] = {
+    val sql: SqlAction[Option[PlayRow], NoStream, Effect.Read] = Plays.sortBy(_.createdAt.desc).take(1).result.headOption
```
Comment (Cardu) [inline src/main/scala/rps/GameRepository.scala]: @veej what about now? I changed the variables to have more straightforward names and explicit type annotation; I think it should be more clear to read and understand

Comment (veej) [reply src/main/scala/rps/GameRepository.scala]: Could this simply be `DBIO[Option[PlayRow]]`?

Comment (Cardu) [reply src/main/scala/rps/GameRepository.scala]: ✅

**Review by veej — APPROVED** [review]
👏
